e2e/terraform/consul-servers.tf

```diff
@@ -0,0 +1,178 @@
+  content = templatefile("etc/consul.d/.environment", {
+    license = var.consul_license
+  })
+  filename        = "uploads/shared/consul.d/.environment"
+  file_permission = "0600"
+}
+
+resource "null_resource" "upload_consul_server_configs" {
+
+  depends_on = [
+    local_sensitive_file.ca_cert,
+    local_sensitive_file.consul_server_config_file,
+    local_sensitive_file.consul_server_key,
+    local_sensitive_file.consul_server_cert,
+    local_sensitive_file.consul_environment,
+  ]
+
+  connection {
+    type            = "ssh"
+    user            = "ubuntu"
+    host            = aws_instance.consul_server.public_ip
+    port            = 22
+    private_key     = file("${path.root}/keys/${local.random_name}.pem")
+    target_platform = "unix"
+    timeout         = "15m"
+  }
+
+  provisioner "file" {
+    source      = "keys/tls_ca.crt"
+    destination = "/tmp/consul_ca.pem"
+  }
+  provisioner "file" {
+    source      = "uploads/shared/consul.d/.environment"
+    destination = "/tmp/.consul_environment"
+  }
+  provisioner "file" {
+    source      = "uploads/shared/consul.d/server_cert.pem"
+    destination = "/tmp/consul_cert.pem"
+  }
+  provisioner "file" {
+    source      = "uploads/shared/consul.d/server_cert.key.pem"
+    destination = "/tmp/consul_cert.key.pem"
+  }
+  provisioner "file" {
+    source      = "uploads/shared/consul.d/servers.hcl"
+    destination = "/tmp/consul_server.hcl"
+  }
+  provisioner "file" {
+    source      = "etc/consul.d/consul-server.service"
+    destination = "/tmp/consul.service"
+  }
+}
+
+resource "null_resource" "install_consul_server_configs" {
+
+  depends_on = [
+    null_resource.upload_consul_server_configs,
+  ]
+
+  connection {
+    type            = "ssh"
+    user            = "ubuntu"
+    host            = aws_instance.consul_server.public_ip
+    port            = 22
+    private_key     = file("${path.root}/keys/${local.random_name}.pem")
+    target_platform = "unix"
+    timeout         = "15m"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "sudo rm -rf /etc/consul.d/*",
+      "sudo mkdir -p /etc/consul.d/bootstrap",
+      "sudo mv /tmp/consul_ca.pem /etc/consul.d/ca.pem",
+      "sudo mv /tmp/consul_cert.pem /etc/consul.d/cert.pem",
+      "sudo mv /tmp/consul_cert.key.pem /etc/consul.d/cert.key.pem",
+      "sudo mv /tmp/consul_server.hcl /etc/consul.d/consul.hcl",
+      "sudo mv /tmp/consul.service /etc/systemd/system/consul.service",
+      "sudo mv /tmp/.consul_environment /etc/consul.d/.environment",
+      "sudo systemctl daemon-reload",
+      "sudo systemctl enable consul",
+      "sudo systemctl restart consul",
+    ]
+  }
+}
+
+# Bootstrapping Consul ACLs:
+#
+# We can't both bootstrap the ACLs and use the Consul TF provider's
+# resource.consul_acl_token in the same Terraform run, because there's no way to
+# get the management token into the provider's environment after we bootstrap,
```

Comment:
Funny thing is we already do have separate TF runs, because we use TF to get the Vault token. So we could probably split this all out into:
* TF for Vault token
* TF to deploy the infra and upload all the config files
* TF to run the bootstrap ACL scripts for Consul and Nomad

I'd definitely like to make this less miserable, but we're already doing the same thing for Nomad ACLs. So let's revisit that under a separate PR.

Reply:
Hum....I haven't tested it yet, but it doesn't seem like we use `consul_*` resources before ACLs are bootstrapped, so I think that we could use a `data "http"` to wait until the Consul agent is fully bootstrapped and have `provisioner "consul"` depend on it so that any Consul resource is only attempted to be created once the provisioner has been configured with the token.

A few "new" toys that could help here would be 
* https://registry.terraform.io/providers/hashicorp/external/latest/docs/data-sources/external
* https://registry.terraform.io/providers/hashicorp/http/latest/docs/data-sources/http

Another non-Terraform option would be to pre-generate a bootstrap token and place it in the [`acl.initial_management`](https://developer.hashicorp.com/consul/docs/agent/config/config-files#acl_tokens_initial_management) config.

But probably not worth refactoring all of this 😅

Reply:
> so I think that we could use a data `"http"`

TIL about `data.http`. Could be handy for sure.

> Another non-Terraform option would be to pre-generate a bootstrap token and place it in the

We're doing that in this PR, but in my testing I still saw "ACLs haven't been bootstrapped" errors. Maybe I'm doing something wrong though.

Reply:
I just remembered an example where I used `data.http` to wait for Nomad:
https://gist.github.com/lgfa29/b707d56ace871602cb4955df2a1afad0#file-main-tf-L137-L144

Configuring the provider with the request output allows every other `nomad_*` resource to wait until everything is ready.
https://gist.github.com/lgfa29/b707d56ace871602cb4955df2a1afad0#file-main-tf-L5-L8

Reply:
Yeah I think there's enough here where we could eliminate these two shell scripts. The `data.http` for Nomad might have to get a little creative inasmuch as we can't give it a client TLS cert, but I think we probably want to eliminate the client verification on HTTP TLS anyways so that the cluster is in line with our current security recommendations.

I've got a small pile of minor refactors I want to tackle once I've got my immediate yak shave done. I'll add this to that pile.